core/tests/test-dependencies/test-to-modules-matcher.ts

```diff
@@ -0,0 +1,234 @@
+  }
+
+  const segments: UrlSegment[] = urlTree.root.children.primary.segments;
+  const segmentGroup: UrlSegmentGroup = urlTree.root.children.primary;
+
+  if (route.path === undefined) {
+    return false;
+  }
+
+  const parts = route.path.split('/');
+  if (parts.length > segments.length) {
+    return false;
+  }
+
+  if (
+    route.pathMatch === 'full' &&
+    (segmentGroup.hasChildren() || parts.length < segments.length)
+  ) {
+    return false;
+  }
+
+  for (let index = 0; index < parts.length; index++) {
+    const part = parts[index];
+    const segment = segments[index];
+    const partIsParameter = part.startsWith(':');
+    if (partIsParameter) {
+      continue;
+    } else if (part !== segment.path) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+export class TestToModulesMatcher {
+  static goldenFilePath: string;
+  static routeToModuleMapping: Map<Route, string> = getRouteToModuleMapping();
+  static collectedModules: string[] = [];
+  static collectedErrors: string[] = [];
+
+  /**
+   * Sets the path of the golden file that contains the list of Angular modules
+   * that are expected to be used by tests.
+   */
+  public static setGoldenFilePath(filePath: string): void {
+    if (!filePath.endsWith('.txt')) {
+      throw new Error('The golden file path must have a .txt extension.');
+    }
+    this.goldenFilePath = filePath;
+  }
+
+  /**
+   * Registers a URL and matches it with an Angular module.
+   */
+  public static registerUrl(url: string): void {
+    if (!url.startsWith(LOCALHOST_URL)) {
+      return;
+    }
+    const path = url.replace(LOCALHOST_URL, '');
+    let matched = false;
+    for (const [route, module] of this.routeToModuleMapping.entries()) {
+      if (!matchUrl(path, route)) {
+        continue;
+      }
+      matched = true;
+      if (this.collectedModules.includes(module)) {
+        break;
+      }
+      if (COMMON_MODULES_TO_EXCLUDE[module]) {
+        let exclude = true;
+        for (const globPattern of COMMON_MODULES_TO_EXCLUDE[module]) {
+          const globMatcher = new glob.GlobSync(globPattern);
+          exclude = !globMatcher.minimatch.match(this.goldenFilePath);
+          if (!exclude) {
+            break;
+          }
+        }
+
+        if (exclude) {
+          break;
+        }
+      }
+      this.collectedModules.push(module);
+      break;
+    }
+
+    if (!matched) {
+      const errorMessage =
+        `No Angular module found for URL: ${url}. Please ensure ` +
+        'that the URL that is being navigated to is a valid Angular route. If it is ' +
+        'a valid Angular route, please ensure that it is being correctly captured by the ' +
+        'core/tests/test-dependencies/route-to-module-mapping-generator.ts script by ensuring ' +
+        'that the URL is in a routing module or is manually mapped. For more information, ' +
+        'please refer to the documentation here: https://github.com/oppia/oppia/wiki/Partial-CI-Tests-Structure.';
```

Comment:
It doesn't look like this wiki page exists yet. Please make sure it is available before merging this PR